### DISCUSSION TITLE
[ideas] if cue -> never do not check the existing targets before skip them

### DISCUSSION THREAD

**Discussion by stemangiola** [ropensci/targets#1357] [Ideas]
### Help

- [X] I understand and agree to <https://books.ropensci.org/targets/help.html>.

### Description

For large pipelines (e.e. 10K targets), I use cue -> never. Since if a target exist by definition should not be recomputed, it would be nice to avoid any check for the skipping of targets.

Currently skipping 40K targets takes many minutes, this makes the pipeline cumbersome, as often has to be restarted for different reasons.

Thanks!

**Comment by wlandau** [comment]
40k targets is a lot. Have you considered batching? (c.f. `tar_rep()` and `tar_map_rep()` from `tarchetypes`)

I don't have plans to change the behavior of cues. Even if `mode = "never"`, we still at least need to check if the target exists. Otherwise, downstream processing is not possible in the general case.

(If you know your upstream targets are up to date, you might have a look at the `shortcut` argument of `tar_make()`, but it might not work well with dynamic branching.)

**Reply by stemangiola** [reply]
> Even if mode = "never", we still at least need to check if the target exists

This check is fine, but I imagine that checking that hundreds of thousand of files exist is incredibly quick (seconds, with `dir()` ?).

In my pipelines, if they are big, it means that they are computationally intensive anyway (each target takes minutes to hours), so in execution mode, I don't mind the overhead of executing targets. However, if I have to restart the pipeline with `cue = "never"` waiting minutes is painful, whereas checking if files exist alone would be very quick.

I understand that tricks can be used to lower the target number, but making the pipeline more complex is not always an amazing solution (targets is great because is lean), so speeding up some simple behaviours would be a win-win.

**Comment by stemangiola** [comment]
From the documentation, I'm not sure if I understand how to convert this

```r
      tar_target(
        name = target_output, 
        command = my_fx(a),
        pattern = map(slice(a, c(1, 2, 3, 5))),
        iteration = "list"
      )
```

to (?)

```r
...
```

**Reply by wlandau** [reply]
Batching is the general concept of grouping multiple tasks into a single target. For example, this simple pipeline processes data one row at a time (non-batched):

```r
library(targets)
list(
  tar_target(name = data, command = airquality),
  tar_target(
    name = analysis,
    command =  tibble::tibble(
      result = log(data$Wind) + log(data$Temp),
      iteration_name = basename(tempfile(pattern = "batch_"))
    ),
    pattern = map(data)
  )
)
```

Here is a batched version which creates a batch for each month and loops over records within each batch.

```r
list(
  tarchetypes::tar_group_by(name = data, command = airquality, Month),
  tar_target(
    name = analysis,
    command =  tibble::tibble(
      result = purrr::map_dbl(seq_len(nrow(data)), ~ log(data$Wind[.x]) + log(data$Temp[.x])),
      batch_name = basename(tempfile(pattern = "batch_"))
    ),
    pattern = map(data)
  )
)
```

Batching lets you find the right tradeoff between the overhead of many targets vs enough targets for parallel processing.

**Reply by stemangiola** [reply]
Thanks, my input targets (e.g. `a` in my example, and `data` in yours) are always lists, is there a simple conversion of my code with `tar_rep `?

**Reply by wlandau** [reply]
`tar_rep()` works with lists, though it's usually cleaner to return tibbles with the main output data and metadata to label it.

**Reply by stemangiola** [reply]
Just to understand. If I group 100 targets in one. What happens if one of the 100 elements fail either for memory or for a stop() command? Does the whole grouped target fail? And if I use `tar_workspace`, does my session tries to load 100 datasets?

**Reply by wlandau** [reply]
> What happens if one of the 100 elements fail either for memory or for a stop() command? Does the whole grouped target fail? 

Yes, it's a tradeoff.

> And if I use tar_workspace, does my session tries to load 100 datasets?

It tries to load the *dependencies* of that batch of 100 datasets, not the datasets that would result from running that target.
